apps/aecore/lib/aecore/oracle/oracle.ex

```diff
@@ -0,0 +1,293 @@
+      )
+
+    {:ok, tx} = SignedTx.sign_tx(tx_data, Wallet.get_private_key())
+    Pool.add_transaction(tx)
+  end
+
+  @doc """
+  Creates a query transaction with the given oracle address, data query
+  and a TTL of the query and response.
+  """
+  @spec query(Account.pubkey(), any(), non_neg_integer(), non_neg_integer(), ttl(), ttl()) ::
+          :ok | :error
+  def query(oracle_address, query_data, query_fee, fee, query_ttl, response_ttl) do
+    payload = %{
+      oracle_address: oracle_address,
+      query_data: query_data,
+      query_fee: query_fee,
+      query_ttl: query_ttl,
+      response_ttl: response_ttl
+    }
+
+    tx_data =
+      DataTx.init(
+        OracleQueryTxData,
+        payload,
+        Wallet.get_public_key(),
+        fee,
+        Chain.lowest_valid_nonce()
+      )
+
+    {:ok, tx} = SignedTx.sign_tx(tx_data, Wallet.get_private_key())
+    Pool.add_transaction(tx)
+  end
+
+  @doc """
+  Creates an oracle response transaction with the query referenced by its
+  transaction hash and the data of the response.
+  """
+  @spec respond(binary(), any(), non_neg_integer()) :: :ok | :error
+  def respond(query_id, response, fee) do
+    payload = %{
+      query_id: query_id,
+      response: response
+    }
+
+    tx_data =
+      DataTx.init(
+        OracleResponseTxData,
+        payload,
+        Wallet.get_public_key(),
+        fee,
+        Chain.lowest_valid_nonce()
+      )
+
+    {:ok, tx} = SignedTx.sign_tx(tx_data, Wallet.get_private_key())
+    Pool.add_transaction(tx)
+  end
+
+  @spec extend(non_neg_integer(), non_neg_integer()) :: :ok | :error
+  def extend(ttl, fee) do
+    payload = %{
+      ttl: ttl
+    }
+
+    tx_data =
+      DataTx.init(
+        OracleExtendTxData,
+        payload,
+        Wallet.get_public_key(),
+        fee,
+        Chain.lowest_valid_nonce()
+      )
+
+    {:ok, tx} = SignedTx.sign_tx(tx_data, Wallet.get_private_key())
+    Pool.add_transaction(tx)
+  end
+
+  @spec data_valid?(map(), map()) :: true | false
+  def data_valid?(format, data) do
+    schema = ExJsonSchema.Schema.resolve(format)
+
+    case ExJsonSchema.Validator.validate(schema, data) do
+      :ok ->
+        true
+
+      {:error, [{message, _}]} ->
+        Logger.error(fn -> message end)
+        false
+    end
+  end
+
+  @spec calculate_absolute_ttl(ttl(), non_neg_integer()) :: non_neg_integer()
+  def calculate_absolute_ttl(%{ttl: ttl, type: type}, block_height_tx_included) do
+    case type do
+      :absolute ->
+        ttl
+
+      :relative ->
+        ttl + block_height_tx_included
+    end
+  end
+
+  @spec calculate_relative_ttl(%{ttl: non_neg_integer(), type: :absolute}, non_neg_integer()) ::
+          non_neg_integer()
+  def calculate_relative_ttl(%{ttl: ttl, type: :absolute}, block_height) do
+    ttl - block_height
+  end
+
+  @spec tx_ttl_is_valid?(SignedTx.t(), non_neg_integer()) :: boolean
+  def tx_ttl_is_valid?(tx, block_height) do
+    case tx do
```

Comment:
Here you say in the `@spec` that tx is `SignedTx.t()` and then in the case you check whether it is some kind of Oracle tx. Am I not understanding the logic here, or the `@spec` is wrong?

Reply:
spec is wrong